designs/dotnetcore-debugging.md

```diff
@@ -0,0 +1,233 @@
+
+And let's break it down to the checklist
+
+* Get and build `vsdbg` for the Lambda container.
+* Mount `vsdbg` to the running Docker container.
+* Add _"wait for the debugger"_ flag to the runner program and implement corresponding functionality.
+* Implement corresponding container entry point override in SAM CLI to support invoking the runner with this flag **on**.
+* Provide instructions for the customer of how to configure VS Code and VS to attach.
+
+### What will be changed
+
+Runner program for .NET core [2.0](https://github.com/lambci/docker-lambda/blob/master/dotnetcore2.0/run/MockBootstraps/Program.cs) and [2.1](https://github.com/lambci/docker-lambda/blob/master/dotnetcore2.1/run/MockBootstraps/Program.cs) in another [repo](https://github.com/lambci/docker-lambda) to support *waiting for the debugger* to attach. I've filed [PR](https://github.com/lambci/docker-lambda/pull/130) with only required changes already (guys and @mhart - reviewers wanted 😁). 
+
+The good part is that no new commands or parameters on SAM CLI side are required.
+
+### Success Criteria
+
+1. Provide handy script to get the `vsdbg` via downloading it to the .NET Lambda runtime container, installing it and getting through the mounting (or manual instructions to achieve the above results). After that user would be able to provide it to SAM CLI via `--debugger-path`.
+2. Provide ready-to-go `launch.json` configuration for attaching to the Lambda function via VS Code.
+3. Customer should be able to easily debug .NET Core 2.0 and 2.1 apps on VS Code (minimal requirement).
+
+### Out-of-Scope
+
+Visual Studio 2017 (can be a target, but it has some other very different Docker container debugging support - needs additional investigation). But some trick with `launch.json` [might be possible](https://github.com/Microsoft/MIEngine/wiki/Offroad-Debugging-of-.NET-Core-on-Linux---OSX-from-Visual-Studio). Help on this area is highly appreciated.
+
+Rider support also needs separate investigation, as it does not support `vsdbg` by any means (licensing [issue](https://github.com/dotnet/core/issues/505)) and therefore they have [their own](https://blog.jetbrains.com/dotnet/2017/02/23/rider-eap-18-coreclr-debugging-back-windows/) debugger implementation and possibly UX around .NET Core remote debugging. If support is required - I think we should open another issue. 
```

Comment:
Rider may be using a ssh based transport, so we need to be extensible of adding that.

Reply:
The worst thing, is that they are not using `vsdbg` 😞 because of the licensing

Reply:
aaaah. :|